### PR TITLE
feat(modifier-key): CapsLock and NumLock state

### PR DESCRIPTION
**Pull request by j10ccc**
## Progress Report

I've implemented this feature and tested it in Safari, Chrome, Firefox on macOS. The modifier key CapsLock works well. However NumLock cannot be treat as a modifier like CapsLock in my laptop. I checked the key in *Keyboard Event Viewer* you recommend in a windows computer, it appears reasonable as the [document](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/getModifierState#modifier_keys_on_firefox) says. So I infer NumLock works fine in windows.

## Implementation

The Keydown event handler in practice page and the `useDepressedKeys` hook in setting page were changed.

Browsers can read modifier key state from OS only by getting a querying function `getModifierState` from `KeyboardEvent`.

In practice page Presenter related logic, I see some code transfer an original `KeyboardEvent` to a custom `KeyEvent` which return a new object contains some basic type properties. I needs call the `getModifierState` from original event, so I just assertion the type and use `defineProperty` to write a readonly object in some cases when updating properties is required.

### Something Compromise

Besides emulating a keyboard event, there is no reasonable approach to **get initial modifier keys state when page loaded or component mounted**. I believe emulating is a hacking way which may affect the user's practice course. So this PR cannot sync OS modifier keys state when page load, but when user press a specific modifier key, its state will be sync.

Also I didn't update states of all modifier keys by pressing any one of them, just the state of the pressed key will be sync. For fully update, it needs a few changes.

### Something Should Be Updated

I declared the modifier keys constant in two places, maybe it is wisely to define such a property to a key when the `loadKeyboard` and `makeKeys` function called? If it is, I could get keyboard keys from the context.

### CODE REVIEW

**Comment by aradzie** [comment]
Wow you put a lot of effort into writing explaining your work!

Yes, I know that the current modifier state can only be obtained from a key press event. So we must wait before the user hits the first key and only then we can display the modifier state.

File: packages/keybr-textinput-events/lib/TextEvents.tsx [hunk 1]
```diff
@@ -226,28 +226,17 @@ export class TextEvents extends PureComponent<Props> {
...
-    location,
-    repeat,
-  };
+/**
+ * Process keyboard event by adding timestamp.
+ * @param ev
+ * @param timeStamp
+ * @returns Original keyboard event with a timestamp
+ */
+function toKeyEvent(ev: KeyboardEvent, timeStamp: number): KeyEvent {
```
Comment (aradzie) [inline packages/keybr-textinput-events/lib/TextEvents.tsx:235]: I would leave the original `KeyEvent` interface. Then I would extend it with a property for the modifiers. Smth like this:

```
interface KeyEvent {
  // A copy of the KeyboardEvent properties.
  // ...
  /** Event occurrence time. */
  readonly timeStamp: number;
  /** The state of keyboard modifier keys. */
  readonly modifiers: readonly string[];
}
```

The `modifiers` property would keep an array of strings such as `"CapsLock"`, `"NumLock"`.

Such a design will make the remaining code below simpler.

File: packages/keybr-textinput-events/lib/types.ts [hunk 2]
```diff
@@ -21,4 +21,5 @@ export type KeyEvent = {
   readonly metaKey: boolean;
   readonly location: number;
   readonly repeat: boolean;
+  readonly getModifierState: (keyArg: string) => boolean;
```
Comment (aradzie) [inline packages/keybr-textinput-events/lib/types.ts:24]: Again, as I explained above, this should be a static array of strings:

```
  /** The state of keyboard modifier keys. */
  readonly modifiers: readonly string[];
```

There are only two modifiers that we care about, "CapsLock" and "NumLock", so a string array would work just fine.

File: packages/page-practice/lib/practice/Presenter.tsx [hunk 3]
```diff
@@ -166,10 +166,21 @@ export class Presenter extends PureComponent<Props, State> {
   };
 
   private handleKeyDown = (ev: KeyEvent): void => {
+    const modifierKeys = ["CapsLock", "NumLock"];
```
Comment (aradzie) [inline packages/page-practice/lib/practice/Presenter.tsx:169]: This line should be moved to where we create a `KeyEvent` instance from a `KeyboardEvent`.

File: packages/page-practice/lib/settings/KeyboardSettings.tsx [hunk 4]
```diff
@@ -146,10 +146,24 @@ function KeyboardPreview(): ReactNode {
   );
 }
 
+/**
+ * Handle modifier keys only on keydown event for os and
+ * browser compatibility.
+ */
 function useDepressedKeys(): readonly string[] {
   const [depressedKeys, setDepressedKeys] = useState<readonly string[]>([]);
+  const modifierKeys = useRef(["CapsLock", "NumLock"]);
```
Comment (aradzie) [inline packages/page-practice/lib/settings/KeyboardSettings.tsx:155]: The same comment.
When we create a `KeyEvent` instance from a `KeyboardEvent`, read the modifier state for these two keys and fill in the `KeyEvent.modifiers` array.

**Review by aradzie — CHANGES_REQUESTED** [review]

**Comment by j10ccc** [comment]
Good idea. I've tested modifier key's behaviour in those three operating systems. This PR still needs some work for compatity. I will follow your advice and solve the problem. It will take some time.

**Comment by j10ccc** [comment]
Is there an easy way to expose the website `http://localhost:3000` to other devices in my network? I try to visit it by using my development laptop ip address, but the browser redirects the page to `http://localhost:3000` which not really exists in the testing device.

Before this, I use Vite to bundle React app and start a dev server. But it seems to that this project uses your own work Fastr to do this. I cannot find related document to resolve it.

**Comment by aradzie** [comment]
> Is there an easy way to expose the website `http://localhost:3000` to other devices in my network? I try to visit it by using my development laptop ip address, but the browser redirects the page to `http://localhost:3000` which not really exists in the testing device.
> 
> Before this, I use Vite to bundle React app and start a dev server. But it seems to that this project uses your own work Fastr to do this. I cannot find related document to resolve it.

If you need this to test the application in different browsers and OSes, don't bother too much. We can test the compatibility afterwards. I have a special domain keybr.xyz for tests and experiments.

**Comment by j10ccc** [comment]
I finished and tested in macOS and Linux. In macOS, the state if modifiers updates **before the key down event**, I can get new state in key down event. In Linux, it is the state updates **after key down and before key up event**. I always get `true` state from key down event after pressing thousand times. 😢 So I have to judge UA to handle different key event behaviors in OSes.

**Comment by aradzie** [comment]
Thanks for your PR!

**Comment by j10ccc** [comment]
I notice that an action runs failed. Some unit test cases I forget update 😢

**Comment by aradzie** [comment]
> I notice that an action runs failed. Some unit test cases I forget update 😢

Don't worry, I'll fix the tests in a minute

**Comment by aradzie** [comment]
I have uploaded your PR to https://www.keybr.xyz/

**Comment by j10ccc** [comment]
Is it possible to use GitHub action to verify the compatibility across OSes by running testes? I know GitHub workflow can declare a specific system to run scripts.

**Comment by aradzie** [comment]
Yes you can run CI actions in different OSes but it's really not worth it.

**Comment by aradzie** [comment]
I wanted to make a pass over your code to clean it up a little bit, remove code duplication, then ideas started flowing, and ended with making a complete rewrite ;) In the process, I rewrote some of my old code as well.

The summary of my changes:

1. I update the keyboard Key component to display the toggled state as a green dot. So a key now can display two different states: key down and key up, toggle off and toggle on.
2. I extracted a global key state listener `ModifierState`. It install its own keyboard event listeners to detect the modifier state, and works independently from the text input.

I removed some of your code but I will restore it a bit later here

https://github.com/aradzie/keybr.com/commit/da02cb4d4a54f42ba2f225fd5247a9aa25b89f95#diff-3938e7ac57e90e33aafb9409c4a868d35c8a197bf25d825051b31d3304105e5dR40

**Comment by j10ccc** [comment]
Good idea. Dealing with a modifier key specially in key down event makes things complex. Using an extra layer to display toggle state is a key to resolve this.🤙

**Pull request merged**